Add permission checker shortcuts to QuarkusSecurityIdentity.Builder

sberyozkin opened the pull request:
Fixes #43717.

This PR adds a few `QuarkusSecurityIdentity.Builder` permission checker shortcuts for users be able to avoid having to write permission checker functions.

You can see the impact on the changed OIDC code which converts scopes to permissions. This code is tested in the oidc-tenancy tests (see #36361).

IMHO we should not be talking about permission checkers in the Permission docs to avoid the confusion, and to try to give a message it is all very simple. We can add some notes later if some users find that simply adding permissions to the identity builder is not enough for them

sberyozkin committed:
Add permission checker shortcuts to QuarkusSecurityIdentity.builder

michalvavrik commented on extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java:
```diff
@@ -205,6 +207,57 @@ public Builder addAttributes(Map<String, Object> attributes) {
+
+                    @Override
+                    public Uni<Boolean> apply(Permission requiredPermission) {
+                        return Uni.createFrom().item(permission.implies(requiredPermission));
```
Strictly speaking this should be fine because we resolve permission checkers uni in parallel. What I thought would be more efficient, but also more verbose, would be to:

- keep `Set<Permission>` on the `Builder` instance
- when `build()` is called, you create one synchronous permission checker like this:

```
for (Permission possessedPermission : permissions) {
    if (possessedPermission.implies(requiredPermission)) {
        return Uni.createFrom().item(true);
    }
}
return Uni.createFrom().item(false);
```
and you add that checker to other checkers before you build `SecurityIdentity`. It would avoid `Uni`. I don't see cons on such approach, however I think my suggestions have so little pros that you don't need to apply it.

sberyozkin replied:
Sure, added your proposal @michalvavrik, have a look please

michalvavrik replied:
looks good, I can't close this comment thread, please consider this thread resolved

michalvavrik commented on extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java:
```diff
@@ -205,6 +207,57 @@ public Builder addAttributes(Map<String, Object> attributes) {
+        }
+
+        /**
+         * Adds permissions.
```
Please mention it is a string permission. I know it is clear from the fact that it accepts a string or from implementation, but if you add `{@link StringPermission}` then IDE can show me without decompiling code (if I ere Quarkus user). It should take just a gist. Thanks

michalvavrik commented on extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java:
```diff
@@ -258,9 +303,32 @@ public QuarkusSecurityIdentity build() {
             return new QuarkusSecurityIdentity(this);
         }
+
+        public void addPossesedPermissionsChecker() {
```
this should be probably private

sberyozkin replied:
Yeah, IDE made it public and I did not notice

michalvavrik reviewed (approved):
LGTM, thank you, I think this will improve user experience

michalvavrik commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1076,7 +1066,7 @@ import jakarta.ws.rs.Path;
+@Path("/project/{projectName}")
 public class ProjectResource {
 
     @PermissionsAllowed("rename-project") <1>
```
Line 1122 says `TIP: Permission checks run by default on event loops whenever possible.`, as you are already fixing permission documentation, maybe you could fix that as well, please? I think I forgot to drop `whenever possible` as Clement wanted. Sorry.

michalvavrik replied:
We can do it later if you don't want.

sberyozkin replied:
@michalvavrik I've np at all, may be it is worth getting a small doc PR out only for `{projectName}` and this minor update so that we can backport ?

michalvavrik replied:
these 2 requests come from `@PermissionChecker` so they cannot be backported. I don't mind opening small PR, but it will be tiny (probably this one tip). I'm fine anyway.

gsmet commented on extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java:
```diff
@@ -205,6 +208,48 @@ public Builder addAttributes(Map<String, Object> attributes) {
+         * @param permissions The permissions in a String format
+         * @return This builder
+         */
+        public Builder addStringPermissions(Set<String> permissions) {
```
I understand why you add to do that here but it's weird to have the singular and plural versions different.

Maybe `addPermissionAsString()` / `addPermissionsAsString()` would be better?

gsmet reviewed (commented):
I added a very small suggestion.

sberyozkin commented:
Thanks @gsmet @michalvavrik, I've tweaked docs, shortcut method signatures.

I've also added unit tests and while working on them I remembered that a conversion like `'read:all' -> new StringPermission("read:all")` is not correct, as we detected earlier during  the OIDC scope to permission conversion, because it won't work against `@PermissionAllowed("read:all")` where `all` is an action and hence `new StringPermission("read", "all")` is produced...

So I've fixed that and it let me remove all the String to Permission conversion code from the OIDC too...

Should be better now...

michalvavrik commented:
No, I didn't expect that `addPermission(String)` works with actions. You would have to do something like `addPermission(String, String...)` having varargs for actions. I didn't catch that issue with OIDC, I suppose tests would fail if you didn't remove them? Anyway, I think having method that accepts `StringPermission` name and actions seemed like a nice shortcut to me.

michalvavrik commented:
> We already have a convention enforced at the @PermissionAllowed("read:all") level where all becomes an action, so I'm not sure adding another shortcut like addPermissionAsString(String permission, String... actions) really works.

Sure, let's don't do it, I don't mind at all. I just say - I liked it :-). Let's not continue on discussing that, it's ok.

> IMHO JavaDocs make it clear, and users have an option to avoid it with addPermission(new StringPermission("read:all"))

`new StringPermission("read:all")` -> `new StringPermission("read", "all")`**** see https://github.com/quarkusio/quarkus-security/blob/main/src/main/java/io/quarkus/security/StringPermission.java#L20

> What do you think ?

TBH I am lost, I think that last commit won't do. Let me come back to it tonight, I am just preparing on something. I'll come back to it and try to understand changes properly. Maybe I missed something. Thanks

sberyozkin commented:
@michalvavrik Sure there is not rush... 

We add these shortcuts to make it easy for users to have constraints like `@PermissionAllowed("read:all")` enforced.
At the `addPermissionAsString(String)` level users don't deal with `Permission`, it is a String, the same String they see in `@PermissionAllowed("read:all")`. But If we don't split it internally into `new StringPermission("read", "all")` then `@PermissionAllowed("read:all")` will fail and users will think why `addPermissionAsString("read:all")` does not work...

Have a look at the [OIDC test resource](https://github.com/quarkusio/quarkus/blob/main/integration-tests/oidc-tenancy/src/main/java/io/quarkus/it/keycloak/ServiceProtectedResource.java). Users see `read:data`. The tokens scopes are `read:data` [too](https://github.com/quarkusio/quarkus/blob/main/integration-tests/oidc-tenancy/src/test/java/io/quarkus/it/keycloak/ServicePublicKeyTestCase.java#L20). 

Now try to comment out `OidcUtils` code which converts the scope like `read:date` into `new StringPermission("read", "date")` and you'll see a failure...

I wonder if I should just drop the comment that a String like `read:all` represents a permission name with an action, the fact it is converted internally to `new StringPermission("read", "all")` is an implementation detail, what matters to users is that `addPermissionAsString("read:all")` helps to get the `@PermissionAllowed("read:all")` check passed...

sberyozkin commented:
Dropped the comments related to how the String to Permission conversion works and that should avoid the confusion

sberyozkin commented:
However, @michalvavrik, I think `StringPermission` JavaDocs should mention that users must use `new StringPermission("read", "all")` for it to work against `@PermissionAllowed("read:all")`, and `@PermissionAllowed` JavaDocs should mention that  `read:all` would work against a Permission with the `read` name and `all` action, but we can take care of it at the quarkus-security project level...

Take you time please to check things, no rush at all, thanks

michalvavrik commented:
I'll comment & review later (maybe tomorrow), just quick note:

> However, @michalvavrik, I think `StringPermission` JavaDocs should mention that users must use `new StringPermission("read", "all")` for it to work against `@PermissionAllowed("read:all")`, and `@PermissionAllowed` JavaDocs should mention that `read:all` would work against a Permission with the `read` name and `all` action, but we can take care of it at the quarkus-security project level...

See https://github.com/quarkusio/quarkus-security/blob/main/src/main/java/io/quarkus/security/PermissionsAllowed.java#L87 and https://github.com/quarkusio/quarkus-security/blob/main/src/main/java/io/quarkus/security/PermissionsAllowed.java#L94. Don't get any better than that :-) I can put it to the `StringPermission` constructor Javadoc, np.

michalvavrik reviewed (approved):
Not sure if you changed it again, but it looks good now :-D

sberyozkin commented:
@michalvavrik I dropped comments related to informing users that we transform it internally :-).

If that looks good to you now then if we can get one more approval, from Steph or Guillaume, it will be good to go

FroMage reviewed (approved):
Looks much better this way :)

sberyozkin commented:
Thanks @FroMage @michalvavrik @gsmet
